Tests: Prefer MUST() over EXPECT(!.is_error())

BenWiederhake opened the pull request:
Inspired by: https://github.com/SerenityOS/serenity/pull/18710#discussion_r1186892445

Should I add a linter for this? It seems like overkill, so I didn't.

Note that `TestLibCoreSharedSingleProducerCircularQueue.cpp::simple_dequeue` is the only non-trivial change; and makes the code even simpler.

<details><summary><code>sed</code> invocation</summary>

Changed by doing:
```
git ls-files '*.cpp' | xargs sed -i -Ee 's,EXPECT\(!(.+)\.is_error\(\)\),MUST(\1),'
```
(… and then reformatting with clang-format.)

</details>

timschumi commented:
Note that we now also have `TRY_OR_FAIL`, which is probably a good middle way between `MUST` (crashy!) and `EXPECT` (will just continue running and not care, probably crashing when trying to access the value).

nico commented:
I like `MUST`, since it always works (for the other one you can't be in a helper fn iirc), and almost always they don't fail anyways, and when they do I have to debug anyways.

In any case I think this change is better than what we currently have 🙂

BenWiederhake commented:
Changes since last push:
- Fixed some compilation errors in `Tests/LibCore/TestLibCoreStream.cpp`, with the nice side effect of making the code shorter and nicer.

ADKaster commented on Tests/LibSQL/TestSqlValueAndTuple.cpp:
```diff
@@ -1240,7 +1240,7 @@ TEST_CASE(shift_right)
 
         auto result = value1.shift_right(value2);
-        EXPECT(!result.is_error());
+        MUST(result);
```
This doesn't work the way you are expecting, because MUST() will call release_value() on the underlying error type and return it from that expression.

That's why all these tests are failing :^). you would have to assign the result of must to a variable in order to compare it.

BenWiederhake replied:
Indeed!

Given that we already call `.value()` anyway, which in turn does `VERIFY(has<T>());`, I'd like to just remove that line entirely.

"But Ben!" you might say, "If you do that, it will crash in case of an error!"
Yes, but that's the current behavior anyway, so there's no loss in testing power.

BenWiederhake commented:
Changes since last push, changed all these for MUST-correctness, to prevent moving out of an `ErrorOr` that is still in use:
- Tests/AK/TestBase64.cpp
- Tests/AK/TestCircularBuffer.cpp
- Tests/LibCompress/TestDeflate.cpp
- Tests/LibCompress/TestGzip.cpp
- Tests/LibGfx/TestFontHandling.cpp
- Tests/LibJS/test-bytecode-js.cpp
- Tests/LibPDF/TestPDF.cpp
- Tests/LibSQL/TestSqlDatabase.cpp
- Tests/LibSQL/TestSqlExpressionParser.cpp
- Tests/LibSQL/TestSqlStatementExecution.cpp
- Tests/LibSQL/TestSqlStatementParser.cpp
- Tests/LibSQL/TestSqlValueAndTuple.cpp

nico commented on Tests/AK/TestCircularBuffer.cpp:
```diff
@@ -37,7 +34,7 @@ void safe_read(CircularBuffer& buffer, u8 supposed_result)
 void safe_discard(CircularBuffer& buffer, size_t size)
 {
     auto result = buffer.discard(size);
-    EXPECT(!result.is_error());
```
Just `MUST(buffer.discard());` (?) (If so, also elsewhere)

nico commented on Tests/Kernel/TestKernelFilePermissions.cpp:
```diff
@@ -82,7 +82,7 @@ TEST_CASE(test_change_file_location)
-    EXPECT(!suid_path_or_error.is_error());
+    MUST(suid_path_or_error);
 
     auto suid_path = suid_path_or_error.release_value().to_deprecated_string();
```
```
auto suid_path = MUST(FileSystem::read_link(DeprecatedString::formatted("/proc/{}/fd/{}", getpid(), fd));).to_deprecated_string()
```

instead of all the stuff above.

(In general, prefer smaller, high-quality changes over huge rewrites with issues like this, maybe?)

BenWiederhake replied:
Yeah, you're right. I thought it would really be a simple search-and-replace, if I had known in advance that each callsite needs individual treatment I would have approached this differently.

nico commented on Tests/LibC/TestIo.cpp:
```diff
@@ -218,7 +218,7 @@ TEST_CASE(unlink_symlink)
-    EXPECT(!target_or_error.is_error());
+    MUST(target_or_error);
 
     auto target = target_or_error.release_value();
```
See above

fdellwing commented:
Why are we not using `TRY_OR_FAIL` instead of `MUST` again?

BenWiederhake commented:
Changes since last push:
- Rebased
- Checked/changed every single callsite again, now that I understand the move-semantics of `MUST()`

Given that it's a two-to-one vote between you people, I'll now change it to `TRY_OR_FAIL`.

BenWiederhake commented:
Changes since last push:
- Rebased
- Instead of unconditionally using `MUST`, this commit now uses `TRY_OR_FAIL` wherever it makes sense.

AtkinsSJ commented:
This has conflicts now.

BenWiederhake commented:
Changes since last push:
- Rebased (the file `Tests/LibGfx/TestImageDecoder.cpp` was updated by someone else to use `MUST` instead of `TRY_OR_FAIL`, so this PR no longer touches that file.)

ADKaster commented on Tests/LibCore/TestLibCoreStream.cpp:
```diff
@@ -344,21 +293,15 @@ TEST_CASE(local_socket_read)
-            EXPECT(!maybe_read_bytes.is_error());
-            auto read_bytes = maybe_read_bytes.release_value();
+            auto receive_buffer = MUST(ByteBuffer::create_uninitialized(64));
+            auto read_bytes = MUST(client_socket->read_some(receive_buffer));
```
Lines 303, 304 and 296 here are MUST() instead of TRY_OR_FAIL. Is that b/c of the different thread context?

fdellwing replied:
`TRY_OR_FAIL` can not be used inside functions. You would need to change the function to return `ErrorOr<T>`, use `TRY` and `TRY_OR_FAIL` the function call.

BenWiederhake replied:
Jup, exactly what Fabian said. I used `TRY_OR_FAIL` in all possible contexts, and the remaining `MUST`s cannot be replaced by `TRY_OR_FAIL` without also changing the code more invasively (i.e. changing return types).

BenWiederhake replied:
On a side note: This code runs in a different thread, and already calls `EXPECT`, which in turn sets `TestSuite::m_current_test_case_passed`, which is a simple `bool`. In other words, that's a data race, or at least feels dangerously close to one.

ADKaster commented on Tests/LibCore/TestLibCoreSharedSingleProducerCircularQueue.cpp:
```diff
@@ -34,9 +34,8 @@ TEST_CASE(simple_dequeue)
-        auto const element = queue.dequeue();
-        EXPECT(!element.is_error());
-        EXPECT_EQ(element.value(), i);
+        auto const element = MUST(queue.dequeue());
```
This file didn't get any TRY_OR_FAIL treatment (just MUST)

BenWiederhake replied:
Huh! You're right, this one slipped through the cracks. Weird.

BenWiederhake replied:
Ah! Now I remember:
```
In file included from /home/user/workspace/serenity/./AK/DeprecatedString.h:9,
                 from /home/user/workspace/serenity/Userland/Libraries/LibCore/SharedCircularQueue.h:13,
                 from /home/user/workspace/serenity/Tests/LibCore/TestLibCoreSharedSingleProducerCircularQueue.cpp:7:
/home/user/workspace/serenity/./AK/Format.h: In instantiation of 'AK::ErrorOr<void> AK::__format_value(TypeErasedFormatParams&, FormatBuilder&, FormatParser&, const void*) [with T = Core::SharedSingleProducerCircularQueue<int>::QueueStatus]':
/home/user/workspace/serenity/./AK/Format.h:298:133:   required from 'AK::VariadicFormatParams<allow_debug_formatters, Parameters>::VariadicFormatParams(const Parameters& ...) [with AK::AllowDebugOnlyFormatters allow_debug_formatters = AK::Yes; Parameters = {char [65], int, Core::SharedSingleProducerCircularQueue<int>::QueueStatus}]'
/home/user/workspace/serenity/./AK/Format.h:569:72:   required from 'void AK::outln(FILE*, CheckedFormatString<Parameters ...>&&, const Parameters& ...) [with Parameters = {char [65], int, Core::SharedSingleProducerCircularQueue<int>::QueueStatus}; FILE = FILE; CheckedFormatString<Parameters ...> = Format::Detail::CheckedFormatString<char [65], int, Core::SharedSingleProducerCircularQueue<int>::QueueStatus>]'
/home/user/workspace/serenity/./AK/Format.h:596:98:   required from 'void AK::warnln(CheckedFormatString<Parameters ...>&&, const Parameters& ...) [with Parameters = {char [65], int, Core::SharedSingleProducerCircularQueue<int>::QueueStatus}; CheckedFormatString<Parameters ...> = Format::Detail::CheckedFormatString<char [65], int, Core::SharedSingleProducerCircularQueue<int>::QueueStatus>]'
/home/user/workspace/serenity/Tests/LibCore/TestLibCoreSharedSingleProducerCircularQueue.cpp:37:30:   required from here
/home/user/workspace/serenity/./AK/Format.h:288:15: error: 'struct AK::Formatter<Core::SharedSingleProducerCircularQueue<int>::QueueStatus, void>' has no member named 'parse'
  288 |     formatter.parse(params, parser);
      |     ~~~~~~~~~~^~~~~
/home/user/workspace/serenity/./AK/Format.h:289:22: error: 'struct AK::Formatter<Core::SharedSingleProducerCircularQueue<int>::QueueStatus, void>' has no member named 'format'; did you mean 'Formatter'?
  289 |     return formatter.format(builder, *static_cast<T const*>(value));
      |            ~~~~~~~~~~^~~~~~
      |            Formatter
```

Or more succinctly: `TRY_OR_FAIL` would try to `format` a `SharedSingleProducerCircularQueue::QueueStatus`, which is an `enum class` without a formatter, so this fails.

I'll add a TODO here, too.

ADKaster commented on Tests/LibCore/TestLibCoreStream.cpp:
```diff
@@ -459,13 +389,11 @@ TEST_CASE(buffered_small_file_read)
     }
-    EXPECT(!file->can_read_line().is_error());
-    EXPECT(!file->can_read_line().value());
+    EXPECT(!TRY_OR_FAIL(file->can_read_line()));
```
This would read clearer to me as 2 lines, like you did below in the buffered_file_tell_and_seek test.

```
bool can_read_line = TRY_OR_FAIL(file->can_read_line());
EXPECT(!can_read_line);
```

BenWiederhake replied:
Absolutely, thanks!

ADKaster commented on Tests/LibJS/test-bytecode-js.cpp:
```diff
@@ -111,15 +110,13 @@ TEST_CASE(loading_multiple_files)
         auto executable = MUST(JS::Bytecode::Generator::generate(test_file_program));
-        auto result = bytecode_interpreter.run(*executable);
-        EXPECT(!result.is_error());
+        MUST(bytecode_interpreter.run(*executable));
```
to confirm, this file has MUST because we cannot format() a `JS::Script` (or w/e the type is here) ? or are there some missed opportunities for TRY_OR_FAIL

BenWiederhake replied:
Yes, it's due to a missing `Formatter<>` implementation.

This is due to `JS::Completion` and not `JS::Script` (afterall, we try to print the error-value, not the success-value), but yes, that's it. Implementing that Formatter feels like it's outside the scope of this PR.

I'll add a TODO, because you're right, failing would be nicer than crashing.

ADKaster reviewed (commented):
The commit message says that it "leaves untouched the places where MUST could be replaced by TRY_OR_FAIL", but also adds many MUST calls 🤔 . Is it worth *not* adding those must calls in the places where we know we want them to be TRY_OR_FAIL?

BenWiederhake commented:
Changes in the last two pushes:
- Split a `EXPECT(!TRY_OR_FAIL(` into two lines (thanks @ADKaster!)
- Add a TODO for two future `MUST` to `TRY_OR_FAIL` conversions due to missing `Formatter` implementations (thanks @ADKaster!)
- Rebased for convenience

ADKaster reviewed (approved):
This seems like an overall improvement to the tests, but I definitely echo Nico's comment that the next batch of changes like this should probably be more targeted and chunked up.

BenWiederhake commented:
Absolutely! I learned my lesson from this.
